Support creating an empty wallet

Sjors opened the pull request:
Add as new argument (`empty`) to `createwallet` to create an empty wallet.

This can be followed by `sethdseed` to create a clean wallet with a custom seed.

Followup PRs can make this significantly more useful, e.g.
* #15006 adds a password argument to `createwallet`
* `TopUpKeyPool()` should look for imported keys
* allow import of a receive and change descriptor with private keys, which also sets the hd seed (custom BIP32 derivation paths, limiting the address type at the wallet level)
*  `TopUpKeyPool()` should look for public keys (if `WALLET_FLAG_DISABLE_PRIVATE_KEYS` is set)

Sjors commented:
Some context:
* https://github.com/bitcoin/bitcoin/pull/14075#issuecomment-446547958 "Import watch only pubkeys to the keypool if private keys are disabled"
* I had difficulty creating a test for handling PSBT in watch-only wallets, where I wanted to mock the signed PSBT by importing the same private private keys into a second wallet and signing it with that: https://github.com/bitcoin/bitcoin/pull/14912/files

promag commented on doc/release-notes-14936.md:
```diff
@@ -0,0 +1,4 @@
+Miscellaneous RPC changes
+------------
+
+- `createwallet` can now create an empty wallet
```
nit
```
- The RPC `createwallet` has now the optional `empty` argument that allows to create empty wallets.
```

promag commented on test/functional/wallet_createwallet.py:
```diff
@@ -0,0 +1,63 @@
+        w1.importpubkey(w0.getaddressinfo(address1)['pubkey'])
+
+        self.log.info("Test empty creation with private keys disabled.")
+        self.nodes[0].createwallet('w2', True, True)
```
nit, could use named arguments.

Sjors replied:
Good idea, I made `wallet_name` into a named argument as well now.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -187,9 +187,13 @@ CPubKey CWallet::GenerateNewKey(WalletBatch &batch, bool internal)
+    // use HD key derivation if HD was enabled during wallet creation and a seed is present
     if (IsHDEnabled()) {
-        DeriveNewChildKey(batch, metadata, secret, (CanSupportFeature(FEATURE_HD_SPLIT) ? internal : false));
+        if (HasHDSeed()) {
```
nit
```cpp
if (!HasHDSeed()) throw std::runtime_error(std::string(__func__) + ": No HD Seed");

DeriveNewChildKey(batch, metadata, secret, (CanSupportFeature(FEATURE_HD_SPLIT) ? internal : false));
```

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2563,6 +2564,11 @@ static UniValue createwallet(const JSONRPCRequest& request)
         disable_privatekeys = request.params[1].get_bool();
     }
 
+    bool empty = false;
```
nit, `create_empty`.

promag reviewed (commented):
Concept ACK.

hebasto commented:
Concept ACK.

laanwj commented:
Concept ACK

@jonasschnelli is this in line with your idea of supporting hardware wallets?

Sjors commented:
It's definitely in line with _my_ idea of adding hardware wallets :-)

See also #14145.

practicalswift commented on test/functional/wallet_createwallet.py:
```diff
@@ -0,0 +1,63 @@
+        self.log.info("Test disableprivatekeys creation.")
+        self.nodes[0].createwallet('w1', True)
+        w1 = node.get_wallet_rpc('w1')
+        assert_raises_rpc_error(-4,"Error: Private keys are disabled for this wallet", w1.getnewaddress)
```
Missing whitespace after `,` – applies throughout this PR :-)

Sjors replied:
Thanks @practicalswift. What Python linter are you running that Travis isn't?

practicalswift replied:
@Sjors We run `flake8` in Travis but the rule `E231` ("missing whitespace after ','") isn't enabled. Please consider opening a PR fixing that :-)

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1426,11 +1430,20 @@ void CWallet::SetHDChain(const CHDChain& chain, bool memonly)
+    // For older wallets the presence of seed_id indicates if HD is enabled.
+    // Support for empty HD capable wallets was added in 0.18
+    return CanSupportFeature(FEATURE_PRE_SPLIT_KEYPOOL) || !hdChain.seed_id.IsNull();
+}
```
Call `HasHDSeed` instead of doing the same check here

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4032,20 +4052,27 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
             walletInstance->SetWalletFlag(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
+        } else if (create_empty) {
+            // Do not generate a new seed
         } else {
```
Instead of an empty else if block, change the else to be `else if (!create_empty)`.

achow101 commented:
Calling `encryptwallet` on an empty wallet does not work. I'm getting;

    DeriveNewSeed: AddKeyPubKey failed

I would expect either that we can't encrypt an empty wallet (bummer, because then we can't have a wallet encrypted from the very beginning), or that it would encrypt and have a newly generated seed (which is what I think you were trying to do).

Sjors commented:
@achow101 thanks, I see you included that commit in #15006

Addressed @promag's, @practicalswift 's and @achow101's comments.

Sjors commented on test/functional/wallet_createwallet.py:
```diff
@@ -0,0 +1,63 @@
+        address1 = w0.getnewaddress()
+
+        self.log.info("Test disableprivatekeys creation.")
+        self.nodes[0].createwallet(wallet_name='w1', disable_private_keys='true')
```
@jnewbery this test fails in `--usecli` if I use a boolean `True` rather than a string `'true'`. Is that by design?

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3979,20 +3996,25 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
-        walletInstance->SetMinVersion(FEATURE_LATEST);
+        // ensure this wallet can only be opened by clients supporting HD with chain split and expects no default key
+        if (!create_empty) {
+            walletInstance->SetMinVersion(FEATURE_PRE_SPLIT_KEYPOOL);
```
@Empact in an earlier version I added a new feature version for this, but then I realized it wasn't necessary. Forgot to remove this bit. Fixed.

achow101 replied:
I think instead of using a version number you should use a wallet flag that is required. The flag can be unset once the wallet is no longer empty. This would remove the need for a new version number while still preventing older software from opening the wallet.

Sjors committed:
QA: Fix -usecli converting named bool args to non-lowercase

Sjors committed:
[wallet] support creating an empty wallet

Sjors commented:
Travis is happy now.

Added `LOCK(cs_wallet)` to `IsHDEnabled` because it uses `CanSupportFeature` which has an `AssertLockHeld`.

I added a new wallet version `FEATURE_ALLOW_EMPTY` which is set as the minimum when creating an empty wallet. When creating a non-empty wallet we stick to `FEATURE_PRE_SPLIT_KEYPOOL` as the minimum version.

achow101 commented:
I don't think it is necessary to set different version numbers depending on whether the wallet can be empty. This is effectively reintroducing the optional version stuff that we had done for HD wallets and HD chain split, and reconciling those was a major pain. I don't see why it is necessary to do that again.

achow101 commented:
> What do you mean by "optional version stuff"?

The original HD wallet stuff was optional and the option was determined by the wallet version number. Non-HD wallets had a lower version number than HD wallets. So an optional feature was determined by the version number. But later when a change to the wallet affected both non-HD and HD wallets and required a version number bump, this optional feature thing switched via the version number had to go away, and making it go away was a pain. In general, I don't think that we shouldn't have optional features be indicated by the version number. Since we have wallet flags now, I think that instead optional features should be indicated through the flags, and if a flag is set that we don't know about, we should throw an error (right now they don't throw an error).

> The alternative seems to be to disallow <= v0.17 nodes from opening a v0.18 wallet, even though only a minitory of users need the empty wallet feature.

That certainly wouldn't be ideal, but I don't think that is really that big of a concern. I think it is far more likely for someone who upgraded to 0.18 from 0.17 to then downgrade to 0.17 than for someone who starts with 0.18 (and thus makes a new wallet) to then downgrade to 0.17.

laanwj commented on src/wallet/wallet.cpp:
```diff
@@ -3214,6 +3224,9 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
             return false;
 
+        // If wallet supports HD but no seed is present, do not top up keypool
+        if(IsHDEnabled() && !HasHDSeed()) return true;
```
maybe roll this logic (which is repeated below) into a function `IsEmpty()`?

laanwj commented:
Concept ACK as this is a dependency for #15006

promag commented:
Should this option be exposed in the GUI (after #13100)?

achow101 commented:
@promag I think it should

jonasschnelli commented:
Tested a bit.
If I create an empty `createwallet dummy false true` and generate an address via GUI I get this (invalid) state:
<img width="599" alt="bildschirmfoto 2019-01-03 um 10 03 24" src="https://user-images.githubusercontent.com/178464/50658885-d526f580-0f3e-11e9-9962-a7d1c92484b2.png">

`getnewaddress` gives me `Error: Keypool ran out, please call keypoolrefill first (code -12)` but I would expect that the key pool gets refilled automatically? not?

Sjors commented:
@jonasschnelli an empty wallet does not have an HD seed, so there is no way to refill the keypool. Calling `sethdseed` will add such a seed.

I'm surprised the GUI doesn't handle this correctly, since the keypool could already run out before this change. Normally it just greys out the new address button. I'll try to reproduce.

Sjors commented:
@achow101 wrote inline:

> I think instead of using a version number you should use a wallet flag that is required. The flag can be unset once the wallet is no longer empty. This would remove the need for a new version number while still preventing older software from opening the wallet.

I'm happy with either option. Mandatory flags explicitly say what a wallet can't do, which is arguably better than inferring capabilities from version numbers. Any other takes?

laanwj commented:
> I'm surprised the GUI doesn't handle this correctly, since the keypool could already run out before this change. Normally it just greys out the new address button. I'll try to reproduce.

Right—apparently the GUI doesn't do correct error handling here, it should handle the case where it's impossible to generate an address.

ryanofsky reviewed (approved):
utACK 8ce17c0fb0ea51e8047db3d5a23bef747bbedf3a, but I agree with @achow101 in preferring a new feature flag to a new version number here.

In this case, since it is unlikely there'd ever be a new version of the wallet which dropped support for unset HD seeds, there's little practical difference between using a new flag or a new version. I prefer flags anyway just for sanity when reading code, because flags make it possible to reason about one feature at a time without having to think about the history of the project. The counterargument in favor of preferring versions to flags is that using flags leads to a combinatorial explosion of configurations that can't all be tested and make the project less stable. I think this argument is spurious because you can always disallow combinations of flags you don't want to test, without the need to confuse future maintainers of the code forever with version number comparisons.

achow101 commented:
> I'm surprised the GUI doesn't handle this correctly, since the keypool could already run out before this change. Normally it just greys out the new address button. I'll try to reproduce.

This error is because the grey out condition is that private keys are disable.

Sjors commented:
Closing in favour of #15226.
